lucet-wiggle/tests/wasi.rs

```diff
@@ -0,0 +1,456 @@
+    // executable.
+    crate::init();
+
+    // Temporary directory for outputs.
+    let workdir = TempDir::new().expect("create working directory");
+
+    // Build a C file into a Wasm module. Use the wasi-sdk compiler, but do not use the wasi libc
+    // or the ordinary start files,
+    // which will together expect various aspects of the Wasi trait to actually work. The C file
+    // only imports one function, `args_sizes_get`.
+    let wasm_build = Link::new(&["tests/wasi_guest.c"])
+        .with_cflag("-nostartfiles")
+        .with_link_opt(LinkOpt::NoDefaultEntryPoint)
+        .with_link_opt(LinkOpt::AllowUndefinedAll)
+        .with_link_opt(LinkOpt::ExportAll);
+    let wasm_file = workdir.path().join("out.wasm");
+    wasm_build.link(wasm_file.clone()).expect("link wasm");
+
+    // We used lucet_wiggle to define the hostcall functions, so we must use it to define our
+    // bindings as well. This is a good thing! No more bindings json files to keep in sync with
+    // implementations.
+    let witx_doc = witx::load(&["../wasi/phases/snapshot/witx/wasi_snapshot_preview1.witx"])
+        .expect("load snapshot 1 witx");
+    let bindings = lucet_wiggle_generate::bindings(&witx_doc);
+
+    // Build a shared object with Lucetc:
+    let native_build = Lucetc::new(wasm_file).with_bindings(bindings);
+    let so_file = workdir.path().join("out.so");
+    native_build
+        .shared_object_file(so_file.clone())
+        .expect("build so");
+
+    // Load shared object into this executable.
+    let module = DlModule::load(so_file).expect("load so");
+
+    // Create an instance:
+    let region = MmapRegion::create(1, &Limits::default()).expect("create region");
+    let mut inst = region.new_instance(module).expect("create instance");
+
+    // Define the TestCtx. This gets put into the embed ctx and is usable from the trait method
+    // calls.
+    let input_a: u32 = 123;
+    let input_b: u32 = 567;
+    let test_ctx = TestCtx {
+        a: input_a,
+        b: input_b,
+        times_called: 0,
+    };
+
+    inst.insert_embed_ctx(test_ctx);
+
+    // Call the `sum_of_arg_sizes` func defined in our
+    // C file. It in turn calls `args_sizes_get`, and
+    // returns the sum of the two return values from that
+    // function.
+    let res = inst
+        .run("sum_of_arg_sizes", &[])
+        .expect("run sum_of_arg_sizes")
+        .unwrap_returned();
+
+    // Check that the return value is what we expected:
+    assert_eq!(res.as_u32(), input_a + input_b);
+
+    let tctx = inst
+        .get_embed_ctx::<TestCtx>()
+        .expect("get test ctx")
+        .expect("borrow");
+    // The test ctx `a` and `b` fields should be the same
+    // as they were initialized to:
+    assert_eq!(tctx.a, input_a);
+    assert_eq!(tctx.b, input_b);
+    // The `arg_sizes_get` trait method implementation
+    // should have incremented `times_called` one time.
+    assert_eq!(tctx.times_called, 1);
```

Comment:
Also, I really like that you've added a `times_called`. Nice attention to detail :slightly_smiling_face: